build.rs

```diff
@@ -590,14 +592,15 @@ fn cc(file: &Path, ext: &str, target: &Target, warnings_are_errors: bool,
         let _ = c.flag("-U_FORTIFY_SOURCE");
     }
     if target.os() == "android" {
-        // Define __ANDROID_API__ to the Android API level we want.
-        // Needed for Android NDK Unified Headers, see:
-        // https://android.googlesource.com/platform/ndk/+/master/docs/UnifiedHeaders.md#Supporting-Unified-Headers-in-Your-Build-System
-        if target.arch() == "aarch64" {
-            // Minimum API level where AArch64 is available is 21.
-            let _ = c.define("__ANDROID_API__", Some("21"));
-        } else {
-            let _ = c.define("__ANDROID_API__", Some("18"));
+        // We need __ANDROID_API__ to be defined to the Android API we want.
+        // To check this, we compile a dummy file that errors if it's not what we want.
+        let required_api_level = if target.arch() == "aarch64" { 21 } else { 18 };
+        let mut tempfile = tempfile::NamedTempFile::new().unwrap();
+        write!(tempfile, "#if __ANDROID_API__ >= {} \n#else\n#error Bad\n#endif\nint main() {{}}",
+               required_api_level).unwrap();
+        if let Err(_) = cc::Build::new().warnings(false).file(tempfile.path()).try_expand() {
+            panic!("You need to define __ANDROID_API__ to at least {} for ring to compile",
+                   required_api_level);
```

Comment:
Rather than generate a temporary file to test this, autoconf style, it seems better to just put the minimum version check `#if` logic in crypto/internal.h. Is there a reason you didn't do it that way?

Reply:
I did it this way for the nicer-looking user-friendlier error message, and also to consider defining the constant ourselves automatically if it is not provided by the user.
But I can do it in `crypto/internal.h` if you want.

Reply:
I propose that we keep the code removal in your patch, but drop the code addition. Then, in the build documentation at https://github.com/briansmith/ring/blob/master/BUILDING.md, let's add a section on building for Android that indicates exactly what one must do to CFLAGS w.r.t. `__ANDROID_API__` when building for Android.

In particular, IIUC, the Android SDK already does a check that `__ANDROID_API__` is defined, so our check is somewhat superflous.